---
title: "TE Distribution and Normalizaiton (by class)"
author: "L.S."
date: "2021/3/4"
output:
  pdf_document:
    toc: no
  html_document:
    theme: cerulean
    toc: yes
    number_sections: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup
```{r}
rm(list = ls())
library(reshape2)
options(scipen = 999)
```

## Data prep.
### load the matrix
TE class were sorted by number, and TFs were sorted by looping z-score.
```{r}
load("./nromedpkn_class_matrix_sorted.Rdata")
normpkn_cls_mx[1:6,1:6]

## transform to long data format
dat <- melt(normpkn_cls_mx)
colnames(dat) <- c("class", "tf", "norm")
```
### Load looping z-score
The TFs were grouped into 3 groups according their looping z-score. The absolute value of 0.8 was seen as the "significant point".
```{r}
lp_zsc <- read.csv("../.././../../hs_getloops/intrachr/3.plot/exless_h.all.tf_des-meanz.v3.csv", header = T)

## group
lp_zsc$group <- ifelse(lp_zsc$allmean_z >= 0.8, "high",
                       ifelse(lp_zsc$allmean_z <= -0.8, "low",
                              "mid"))

## reverse the tf order for later heatmap usage
dat$tf <- factor(dat$tf,
                 levels = rev(colnames(normpkn_cls_mx)))

## assign colors to TFs in diff. groups
labelcol <- ifelse(lp_zsc$group == "high", "#CD5C5C",   #'high' = 'IndianRed'
                  ifelse(lp_zsc$group == "mid", "#DAA520", #'mid' = 'Goldenrod'
                         "#4682B4"))  #'low' = 'SteelBlue'
```


## Un-transformed matrix
### Barplot 
Stacked and percent stacked barplot to show the number and proportion.
Setup and color:
```{r}
library(ggplot2)

theme_set(
  theme_minimal() +
    theme(legend.position = "right",
          legend.key.size = unit(12, "pt"))
)

library(ggsci)
simpsons <- pal_simpsons("springfield")(16)
```

#### Stacked barplot
```{r, fig.width = 7, fig.height = 12}
# stack
p <- ggplot(data = dat,
            aes(x = tf, y = norm, fill = class))
p + geom_bar(stat = "identity", position = "stack") +
  labs(title = "TE Class Distribution",
       x = NULL, y = "Normed TE Number") +
  coord_flip() +
  scale_fill_manual(values = simpsons) +
  guides(fill = guide_legend(title = "TE Class", reverse = F)) +
  theme(axis.text.x = element_text(size = 8),
        axis.text.y = element_text(size = 7, colour = rev(labelcol), face = "bold"),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6))
```

#### Percent stacked barplot
```{r, fig.width = 7, fig.height = 12}
# fill
p <- ggplot(data = dat,
            aes(x = tf, y = norm, fill = class))
p + geom_bar(stat = "identity", position = "fill") +
  labs(title = "TE Class Distribution",
       x = NULL,y = "Normed TE Proportion") +
  coord_flip() +
  scale_fill_manual(values = simpsons) +
  guides(fill = guide_legend(title="TE Class", reverse = F)) +
  theme(axis.text.x = element_text(size = 8),
        axis.text.y = element_text(size = 7, colour = rev(labelcol), face = "bold"),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6))
```

### Histogram
```{r}
theme_set(
  theme_classic() +
    theme(legend.position = "none")
)

p <- ggplot(dat, aes(x = norm))
p + geom_histogram(aes(y = stat(density), colour = "black"),
                   fill = "white", bins = 100) +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  scale_color_manual(values = c("#868686FF")) +
  labs(title = "Normalized TE Num Histogram",
       subtitle = "Human",
       caption = "bins = 100") +
  ylab("Dens.") +
  xlab("Normalized TE number")
```

### Heatmap
Bins in histogram might give some hints on how to set the "break" of the color legend of heatmap..?
```{r}
range(normpkn_cls_mx) #resized to 200 bp, back to the peak number
quantile(normpkn_cls_mx)
bk <- seq(min(normpkn_cls_mx), max(normpkn_cls_mx), length.out = 100) #uniform breaks
```

Set the annotation column and color.
```{r}
library(viridis)

anno_tf <- data.frame(lp_zsc$group, row.names = lp_zsc$X)
colnames(anno_tf) <- "looping score"

## anno color
looping <- list(`looping score` = c(high = "#CD5C5C",
                                    mid = "#DAA520",
                                    low = "#4682B4"))
```

#### Heatmap of uniform breaks
```{r, fig.show = "hide"}

grp <- pheatmap::pheatmap(t(normpkn_cls_mx), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = bk,
                   color = viridis(length(bk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap default",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(normpkn_cls_mx), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = bk,
                   color = viridis(length(bk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap default clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```

#### Heatmap of quantile breaks
> https://slowkow.com/notes/pheatmap-tutorial/#quantile-breaks

```{r}
quantile_breaks <- function(xs, n) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}
qbk <- quantile_breaks(normpkn_cls_mx, n = 100)
```

```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(normpkn_cls_mx), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = qbk,
                   color = viridis(length(qbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap default with quantile breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(normpkn_cls_mx), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = qbk,
                   color = viridis(length(qbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap default with quantile breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```

## Z-score normalization
z-score normalized by between TFs.
```{r}
zn_tecls <- t(scale(t(normpkn_cls_mx))) #(n-row.mean)/row.sd
zn_tecls[1:6,1:6]

## long data format for ggplot2 histogram
zdat <- melt(zn_tecls)
colnames(zdat) <- c("class", "tf", "norm")
```

### Histogram
```{r}
theme_set(
  theme_classic() +
    theme(legend.position = "none")
)

p <- ggplot(zdat, aes(x = norm))
p + geom_histogram(aes(y = stat(density), colour = "black"),
                   fill = "white", bins = 50) +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  scale_color_manual(values = c("#868686FF")) +
  labs(title = "Z-Score Normalized TE Num Histogram",
       subtitle = "Human",
       caption = "bins = 50") +
  ylab("Dens.") +
  xlab("Z-Score Normalized TE number")
```

### Heatmap
Set uniform and quantile breaks
```{r}
range(zn_tecls)
quantile(zn_tecls)
zbk <- seq(min(zn_tecls), max(zn_tecls), length.out = 100) #uniform breaks
qzbk <- quantile_breaks(zn_tecls, n = 100) #quantile breaks
```

#### Heatmap of uniform breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(zn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = zbk,
                   color = viridis(length(zbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap z-score normalization with uniform breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(zn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = zbk,
                   color = viridis(length(zbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap z-score normalization with uniform breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```

#### Heatmap of quantile breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(zn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = qzbk,
                   color = viridis(length(qzbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap z-score normalization with quantile breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(zn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = qzbk,
                   color = viridis(length(qzbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap z-score normalization with quantile breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```


## Quantile normalization
```{r}
library(preprocessCore)
qn_tecls <- normalize.quantiles(normpkn_cls_mx)
dimnames(qn_tecls) <- list(row.names(normpkn_cls_mx), colnames(normpkn_cls_mx))
qn_tecls[1:6,1:6]

## long data format for ggplot2 histogram
qdat <- melt(qn_tecls)
colnames(qdat) <- c("class", "tf", "norm")
```

### Histogram
```{r}
theme_set(
  theme_classic() +
    theme(legend.position = "none")
)

p <- ggplot(qdat, aes(x = norm))
p + geom_histogram(aes(y = stat(density), colour = "black"),
                   fill = "white", bins = 18) +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  scale_color_manual(values = c("#868686FF")) +
  labs(title = "Quantile Normalized TE Num Histogram",
       subtitle = "Human",
       caption = "bins = 18") +
  ylab("Dens.") +
  xlab("Quantile Normalized TE number")
```

### Heatmap
Set uniform and quantile breaks
```{r}
range(qn_tecls)
quantile(qn_tecls)
uni_qnbk <- seq(min(qn_tecls), max(qn_tecls), length.out = 18) #uniform breaks
q_qnbk <- quantile_breaks(qn_tecls, n = 18) #quantile breaks
```

#### Heatmap of uniform breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(qn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = uni_qnbk,
                   color = viridis(length(uni_qnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap quantile normalization with uniform breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(qn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = uni_qnbk,
                   color = viridis(length(uni_qnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap quantile normalization with uniform breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```

#### Heatmap of quantile breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(qn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = q_qnbk,
                   color = viridis(length(q_qnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap quantile normalization with quantile breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(qn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = q_qnbk,
                   color = viridis(length(q_qnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap quantile normalization with quantile breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```


## Quantile normalization (normed TE matirx transpose)
In Xiaohua Shen's Cell Reports paper(Lu, J. Yuyang, et al. "Genomic repeats categorize genes with distinct functions for orchestrated regulation." Cell reports 30.10 (2020): 3296-3311.), they have a matrix(40,005 * 84) of 40,005 transcripts and 84(14 * 6) repetive elements(of different genic regions).
> The raw repetitive element matrix (40,005384) was normalized with "normalize.quantiles" function in "pre- processCore" package in R software.

```{r}
library(preprocessCore)
tqn_tecls <- t(normalize.quantiles(t(normpkn_cls_mx)))
dimnames(tqn_tecls) <- list(row.names(normpkn_cls_mx), colnames(normpkn_cls_mx))
tqn_tecls[1:6,1:6]

## long data format for ggplot2 histogram
tqdat <- melt(tqn_tecls)
colnames(tqdat) <- c("class", "tf", "norm")
```

### Histogram
```{r}
theme_set(
  theme_classic() +
    theme(legend.position = "none")
)

p <- ggplot(tqdat, aes(x = norm))
p + geom_histogram(aes(y = stat(density), colour = "black"),
                   fill = "white", bins = 50) +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  scale_color_manual(values = c("#868686FF")) +
  labs(title = "T-Quantile Normalized TE Num Histogram",
       subtitle = "Human",
       caption = "bins = 18") +
  ylab("Dens.") +
  xlab("T-Quantile Normalized TE number")
```

### Heatmap
Set uniform and quantile breaks
```{r}
range(tqn_tecls)
quantile(tqn_tecls)
uni_tqnbk <- seq(min(tqn_tecls), max(tqn_tecls), length.out = 50) #uniform breaks
q_tqnbk <- quantile_breaks(tqn_tecls, n = 50) #quantile breaks
```

#### Heatmap of uniform breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(tqn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = uni_tqnbk,
                   color = viridis(length(uni_tqnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap t-quantile normalization with uniform breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(tqn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = uni_tqnbk,
                   color = viridis(length(uni_tqnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap t-quantile normalization with uniform breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```

#### Heatmap of quantile breaks
```{r, fig.show = "hide"}
grp <- pheatmap::pheatmap(t(tqn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = F, ## Group by looping z-score
                   breaks = q_tqnbk,
                   color = viridis(length(q_tqnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap t-quantile normalization with quantile breaks",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)

clus <- pheatmap::pheatmap(t(tqn_tecls), scale = "none",
                   cluster_cols = T,
                   cluster_rows = T, ## TFs clustered
                   breaks = q_tqnbk,
                   color = viridis(length(q_tqnbk)),
                   border_color = "grey20",
                   clustering_distance_rows = "correlation",
                   main = "pheatmap t-quantile normalization with quantile breaks clustered",
                   cellwidth = 12,
                   cellheight = 12,
                   annotation_row = anno_tf,
                   annotation_colors = looping)
```

```{r, fig.width = 12, fig.height = 20}
require(ggplotify)
p1 <- as.ggplot(grp)
p2 <- as.ggplot(clus)
cowplot::plot_grid(p1, p2, ncol=2)
```